### PR TITLE
chore(build): enable AoT

### PR DESCRIPTION
**Pull request by robwormald**
Enables AoT compilation via the new `@ngtools/webpack` plugin. 

Supersedes #176

### CODE REVIEW

**Comment by DanielJDufour** [comment]
@robwormald , is this still an issue?  If yes, could you explain in more detail what the PR does and why AOT compilation is necessary?

**Comment by robwormald** [comment]
@DanielJDufour it's necessary for sure. 

In a nutshell, Angular takes html-like templates:
```html
<h1>Hello {{name}}</h1>
```
and *compiles* them to javascript code:
```ts
renderer.createElement('h1');
renderer.setText(element, 'Hello ' + name)
```

For development, Angular can do this at *runtime* - this is "Just in Time" (JIT) mode - which is currently how code.gov does. This allows for a quick development cycle when working locally, but incurs a *massive* penalty in terms of both code size and performance. The work of compiling the templates happens *every* time the site loads, and on mobile devices this can take many seconds, in addition to shipping the several hundred KB of the actual compiler code.

By contrast, AOT (Ahead of Time) mode does this compilation work once, at build time. This means the compiler itself does not have to ship to the browser, leading to much less code being shipped, and no waiting for templates to compile.

The context is here: https://github.com/GSA/code-gov-web/issues/129

This PR is probably stale. I'd like to open a new issue and get this solved - its literally the single biggest performance win you could do. If there is interest, we'd love to collaborate on the site with the code.gov team to make it an awesome experience for everyone.

**Comment by DanielJDufour** [comment]
@robwormald , thank you.  That was very helpful and the AOT approach looks very promising.

I haven't used AOT compiling before, so I have some newb questions:
- Is it possible to specify what components are precompiled and which are not?  We're currently in the process of integrating our API into the front-end with the theory being that we can lazy load components to avoid loading data over the network that we don't need.  It will also allow us to avoid cutting a new release every time our data changes.
- Is it possible to AOT compile a template with initial data and then recompile when new data is acquired.  I'm specifically thinking of our search-input on the first page.  Is it possible to AOT compile it then recompile when text is entered and auto-complete results come in via the API.

Thanks again!

**Comment by DanielJDufour** [comment]
FYI, I'm going to be working on integrating your changes into a new branch today.  Thanks!

**Comment by robwormald** [comment]
Great! Do reach out if you need any help!

**Comment by DanielJDufour** [comment]
@robwormald , thanks for the offer.  I actually do have one question if you don't mind.  I looked over your changes and googled "Angular AOT" and came across this page https://angular.io/guide/aot-compiler.  The page seems to suggest that we can implement AOT without a custom plugin (but I might be reading it wrong).  I was wondering if `ng build --prod` is a different option and if it is, what are the pros and cons versus using `@ngtools/webpack`?  Thanks!

File: src/app/app.shell.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { NgModule, ApplicationRef } from '@angular/core'
```
Comment (DanielJDufour) [inline src/app/app.shell.ts]: @robwormald , what does this file do?

**Comment by DanielJDufour** [comment]
FYI: I took your changes from your relevant commits and moved them into a new branch (which is synced up with master).  You can view it here: https://github.com/GSA/code-gov-web/pull/443.

I'm going to close this PR now then and hope we can discuss on the new PR.

**Pull request closed without merging**